src/runtime/transport.ts

```diff
@@ -0,0 +1,190 @@
+import {
+  toConnectedEvent,
+  toMessageEvent,
+  createMessageEvent,
+  toRealTimeMessage,
+  createHeartbeatEvent,
+  createDisconnectEvent,
+  Encoding,
+} from "./messages";
+import { EventEmitter } from "node:events";
+import * as proto from "../proto/server/v1/realtime";
+import { ErrorEvent } from "ws";
+
+type MessageEventListener = (MessageEvent: proto.MessageEvent) => void;
+
+export type Newable<T> = { new (...args: any[]): T };
+
+interface TransportConfig {
+  heartbeatTimeout: number;
+  WebSocket: Newable<WebSocket>;
+}
+
+interface Session {
+  sessionId: string;
+  socketId: string;
+}
+
+type ConnectionState =
+  | "failed"
+  | "connecting"
+  | "connected"
+  | "uninitialized"
+  | "closing"
+  | "closed";
+
+export class Transport extends EventEmitter {
+  private channelListeners: Map<string, MessageEventListener[]>;
+  private ws: WebSocket;
+  private session?: Session;
+  private _isConnected: Promise<void>;
+  private _connectionState: ConnectionState = "uninitialized";
+  private heartbeatId: number | NodeJS.Timeout;
```

Comment:
is this heartbeatTimer? Or heartbeatTimeout?

Reply:
This is the id to the `setTimeout` function so that we can cancel it if we send a message. We only want to fire it when the socket is idle.